### PR TITLE
feat: Added Sweepstake banner

### PR DESCRIPTION
**Pull request by NidhiKJha**
## **Description**

Added a time-sensitive sweepstakes banner to the carousel component that automatically manages its visibility based on campaign dates:

- Banner is only visible between April 9th, 2025 00:00:00 UTC and April 15th, 2025 23:59:59 UTC
- When active, sweepstakes banner appears as the first slide in the carousel
- Fund slide maintains its position as first slide when sweepstakes is inactive, second when active
- Added test coverage:
  - Unit tests for date-based visibility logic
  - Edge case handling for start/end dates
  - Test utilities for simulating different dates

### Implementation Details
- Added `getSweepstakesCampaignActive` function to handle date validation
- Created `useCarouselManagement` hook to manage slide ordering 
- All date comparisons use UTC to ensure consistent behavior across timezones
- Slides maintain proper order and properties during state updates

### Testing Notes
- Unit tests cover all date scenarios (before, during, after campaign)
- Edge cases tested for exact start/end times
- Verified slide ordering and properties in all states

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/31246?quickstart=1)

## **Related issues**

Fixes: [4467](https://github.com/MetaMask/MetaMask-planning/issues/4467)

## **Manual testing steps**
**IMPORTANT**: Make sure to reinstall the app whenever you dismiss the sweepstakes banner. Redux persists the dismissed banner between sesesisons

1. Current date testing:
 - Open the app today (before April 9th, 2025)
 - Verify the sweepstakes banner is not visible
 - Confirm fund slide appears first in the carousel

2. Active period testing:
 - In `ui/hooks/useCarouselManagement/constants.ts`, change the date to be within our current date
  ```typescript
 export const SWEEPSTAKES_START = new Date('2025-03-25T00:00:00Z'); // Date during sweepstakes
 export const SWEEPSTAKES_END = new Date('2025-04-15T23:59:59Z');
 ```
 - Verify sweepstakes banner appears and is the first slide
 - Verify fund slide appears second
 - Verify all other slides maintain correct order
 
3.  Inactive period testing:
 - In `ui/hooks/useCarouselManagement/constants.ts`, change the dates to be past our current date
  ```typescript
 export const SWEEPSTAKES_START = new Date('2025-02-25T00:00:00Z'); // Dates after sweepstakes
 export const SWEEPSTAKES_END = new Date('2025-02-30T23:59:59Z');
 ```
 - Verify sweepstakes banner disappears
 - Verify the fund banner is the first slide
 - Verify all other slides maintain correct order

### **Before**

Screenshot 2025-03-25 at 1 19 00 PM

### **After**

Screenshot 2025-03-25 at 1 24 17 PM

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/controllers/app-state-controller.ts [hunk 1]
```diff
@@ -590,7 +590,7 @@ export class AppStateController extends BaseController<
         );
       });
 
-      state.slides = [...updatedCurrentSlides, ...newSlides];
```
Comment (vinnyhoward) [inline app/scripts/controllers/app-state-controller.ts:593]: Swapped this because new slides would append to the end, since there `updateSlide` action is only used on the `ui/components/multichain/account-overview/account-overview-layout.tsx` it doesn't really matter if its changed

File: ui/components/multichain/account-overview/account-overview-layout.tsx [hunk 2]
```diff
@@ -56,34 +49,14 @@ export const AccountOverviewLayout = ({
...
   const hasZeroBalance = totalBalance === ZERO_BALANCE;
+  const { slides } = useCarouselManagement({
+    hasZeroBalance,
+  });
 
   ///: BEGIN:ONLY_INCLUDE_IF(build-main,build-beta,build-flask)
   const { openBridgeExperience } = useBridging();
   ///: END:ONLY_INCLUDE_IF
 
-  useEffect(() => {
```
Comment (vinnyhoward) [inline ui/components/multichain/account-overview/account-overview-layout.tsx:64]: Refactored the slides logic and created its own hook, that way its easier to test in isolation

File: ui/components/multichain/carousel/carousel.tsx [hunk 3]
```diff
@@ -41,12 +42,26 @@ export const Carousel = React.forwardRef(
     const visibleSlides = slides
       .filter((slide) => !slide.dismissed || slide.undismissable)
       .sort((a, b) => {
+        const isSweepstakesActive = getSweepstakesCampaignActive(
```
Comment (vinnyhoward) [inline ui/components/multichain/carousel/carousel.tsx:45]: This is needed so that the sweepstakes banner is shown first over the fund banner when it is not dismissible

**Review by salimtb — APPROVED** [review]
tested and approved , LGTM

**Review by sahar-fehri — APPROVED** [review]

**Review by EtherWizard33 — APPROVED** [review]
LGTM, approved.

It made sense and was readable, if any changes are require it looks like it will be maintainable.

I understand the timing might be tight to merge this in time, which is probably why dates are in the code. On the flip side this could quickly turn out to be a content management system (CMS) so down the road expectations should be managed about this because those have caveats. That being said, probably having them in a remote feature flag, as a remote config, would allow for any date adjustement and could prove useful to easily adjust the campain dates.
